src/unix/thread.c

```diff
@@ -502,26 +506,138 @@ int uv_sem_trywait(uv_sem_t* sem) {
 
 #else /* !(defined(__APPLE__) && defined(__MACH__)) */
 
-int uv_sem_init(uv_sem_t* sem, unsigned int value) {
+#ifdef __GLIBC__
+
+/* Hack around https://sourceware.org/bugzilla/show_bug.cgi?id=12674
+ * by providing a custom implementation for glibc < 2.21 in terms of other
+ * concurrency primitives.
+ * Refs: https://github.com/nodejs/node/issues/19903 */
+
+/* To preserve ABI compatibility, we treat the uv_sem_t as storage for
+ * a pointer to the actual struct we're using underneath. */
+
+static uv_once_t glibc_version_check_once = UV_ONCE_INIT;
+static int glibc_needs_custom_semaphore = 0;
+
+static void glibc_version_check(void) {
+  const char* version = gnu_get_libc_version();
+  glibc_needs_custom_semaphore =
+      version[0] == '2' && version[1] == '.' &&
+      atoi(version + 2) < 21;
+}
+
+#else /* !defined(__GLIBC__) */
+
+#define glibc_needs_custom_semaphore 0
+
+#endif
+
+typedef struct uv_semaphore_s {
+  uv_mutex_t mutex;
+  uv_cond_t cond;
+  unsigned int value;
```

Comment:
volatile?

Reply:
@gireeshpunathil It’s protected by a mutex, so I think we’re good here?

Reply:
@addaleax - I don't think that alone will help, there is nothing preventing the word being cached in an L2 D-cache of the thread's CPU - that leads to other threads getting stale values. The cache lines are immediately flushed only if the word is a volatile. If you doubt, I can check and confirm - in platforms like PPC that implement weakly consistent memory model, the issue will be pronounced, and we can easily test.

Reply:
@gireeshpunathil They generate the same machine code, byte-for-byte, for me locally… if you are sure that on PPC there’s a difference, I’d be curious about that.

Also, according to [this SO answer](https://stackoverflow.com/a/24138417/688904), the mutex implementation already issues appropriate memory barriers on its own, which matches my expectations and goes beyond what would `volatile` would effect anyway?

Like, if you want, I can add `volatile` – there is probably no harm in doing so. But I’m more or less confident that this is not really it’s purpose.

Reply:
Yes, `volatile` isn't necessary; `uv_mutex_lock()` and `uv_mutex_unlock()` are compiler and memory barriers.

Aside, an architecture's memory model is really only relevant with multiple loads or stores to different locations.  I'm not aware of any architecture where a single load or store on a naturally aligned word boundary isn't atomic.

(Doesn't mean the compiler can't reorder it with other loads and stores but that's a different issue.)

Reply:
I did few tests to convince myself that the mutex applies memory barriers on the data it protects (effect of executing `mfence` / `msync`) in the platforms I tested, though I don't see any documented support for mutex APIs on this behavior. It may be that the mutex's lock variable itself requires fencing to synchronize with other threads, and the fencing on the mutex variable also covers the data?

In terms of memory models: my point was not on atomicity (I agree most of the modern instrcutions are atomic for words) but on visibility: stores made by one CPU on cached data to be made available to other CPUs in the very next CPU cycle, `volatile` is the only option I believe, else it lives in cache, un-flushed. Unfortunately in my tests, I wasn't able to place my word in cache, so my theory stands un-proven. Volatile or non-volatile, code and output are different, but gcc never produced fencing instructions, so I think the compiler knows that this my word will never be cached, and will be written-through.

The below code is mere manifestation of compiler using memory vs. register for the stores, based on volatile or non-volatile, no caching related considerations in the picture at all.


```c
#include <pthread.h>
#include <stdio.h>

int flag;

// word in question
int sem;

void* w(void* arg) {

  //don't escape into OS primitives, stay in the user space.
  while(1) {
    if( sem > (1024 * 1024)) sem = 0;
    sem++;
  }
}
int main() {
  pthread_t thread;
  pthread_create(&thread, NULL, w, NULL);
  while(1) {
    if(flag++ > (1024 * 1024 * 1024)) {
      flag = 0;
      fprintf(stderr, "%d\n", sem);
    }
  }
  pthread_join(thread, NULL);
  return 0;
}
```
$  gcc rec.c -lpthread -O3
$  ./a.out
```
0
0
0
^C
```
with
```patch
- int sem;
+ volatile int sem;
```

$  gcc rec.c -lpthread -O3
$  ./a.out
```
475321
475321
1040652
^C
```

Reply:
> It may be that the mutex's lock variable itself requires fencing to synchronize with other threads, and the fencing on the mutex variable also covers the data?

Yes, pretty much.  You can't really implement a mutex without making it a full barrier; all effects from the critical section should be immediately visible to other threads when the mutex is unlocked.

> `volatile` is the only option I believe, else it lives in cache, un-flushed.

No, `volatile` is exclusively a compiler construct.  For cache coherency, you need memory barriers.

`volatile` tells the compiler that the pointed-to memory may change between loads and stores and that it's not safe to reorder with other volatile loads and stores.  It's still allowed to reorder with non-volatile loads and stores, so you can't use it to delineate critical sections.

That said, a `volatile int` coupled with a `__asm__ __volatile__ ("" ::: "memory")` compiler barrier is sufficient to implement some synchronization primitives on architectures with strong memory models (like x86_64) but you'd still need to be careful of reordering at the hardware level.

(And if you really want to go off the deep end: you can avoid expensive memory barriers on x86 by simply doing something else until the store has propagated to other cores.  The timing is micro-architecture dependent but it's usually > 200 cycles.)